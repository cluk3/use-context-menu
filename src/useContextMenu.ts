import {
  useCallback,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import { getMenuPosition, getRTLMenuPosition } from "./helpers";
import buildUseContextMenuTrigger, {
  UseContextTriggerResult,
} from "./buildUseContextMenuTrigger";
import { Coords, ContextTriggerConfig } from "./index.d";

export enum keyCodes {
  ESCAPE = 27,
  ENTER = 13,
  TAB = 9,
  UP_ARROW = 38,
  DOWN_ARROW = 40,
}
const baseStyles = {
  position: "absolute",
  opacity: 0,
  pointerEvents: "none",
  top: 0,
  left: 0,
  userSelect: "none",
  transform: "translate3d(0,0,0)",
};

const focusElement = (el: HTMLElement): void => el.focus();
const useContextMenu = ({
  rtl = false,
  handleElementSelect = focusElement,
  hideOnScroll = false,
} = {}): UseContextMenuResult => {
  const menuRef = useRef<HTMLElement>();
  const selectables = useRef<HTMLElement[]>([]);
  // TODO: refactor with useReducer
  const [style, setStyles] = useState(baseStyles);
  const [selectedIndex, setSelectedIndex] = useState(-1);
  const [isVisible, setVisible] = useState(false);
  const isVisibleRef = useRef(isVisible);
  isVisibleRef.current = isVisible;
  const [coords, setCoords] = useState<[number, number]>([0, 0]);
  const [collectedData, setCollectedData] = useState<unknown>();

  const hideMenu = useCallback(() => {
    setVisible(false);
    setSelectedIndex(-1);
  }, [setVisible, setSelectedIndex]);

  const triggerVisible = useCallback(
    (coords: Coords, data: unknown) => {
      setVisible(true);
      setCoords(coords);
      setCollectedData(data);
    },
    [setVisible, setCollectedData, setCoords]
  );

  const markSelectable = (el: HTMLElement): HTMLElement[] =>
    (selectables.current = el === null ? [] : [...selectables.current, el]);

  useEffect(() => {
    const handleOutsideClick = (e: MouseEvent | TouchEvent): void => {
      if (
        isVisibleRef.current &&
        menuRef.current &&
        !menuRef.current.contains(e.target as HTMLElement)
      ) {
        hideMenu();
        document.removeEventListener("mousedown", handleOutsideClick);
        document.removeEventListener("touchstart", handleOutsideClick);
      }
    };
    const handleKeyNavigation = (e: KeyboardEvent): void => {
      switch (e.keyCode) {
        case keyCodes.TAB:
          hideMenu();
          break;
        case keyCodes.ESCAPE:
          e.preventDefault();
          hideMenu();
          break;
        case keyCodes.UP_ARROW:
          e.preventDefault();
          if (selectedIndex > 0) {
            setSelectedIndex((s) => s - 1);
            handleElementSelect(selectables.current[selectedIndex - 1]);
          }
          break;
        case keyCodes.DOWN_ARROW:
          e.preventDefault();
          if (selectedIndex + 1 < selectables.current.length) {
            setSelectedIndex((s) => s + 1);
            handleElementSelect(selectables.current[selectedIndex + 1]);
          }
          break;
        case keyCodes.ENTER:
          if (selectedIndex !== -1) {
            selectables.current[selectedIndex].click();
          }
          hideMenu();
          break;
        default:
      }
    };
    if (isVisible) {
      document.addEventListener("mousedown", handleOutsideClick, {
        passive: true,
        capture: true,
      });
      document.addEventListener("touchstart", handleOutsideClick, {
        passive: true,
        capture: true,
      });
      hideOnScroll &&
        document.addEventListener("scroll", hideMenu, {
          once: true,
        });
      document.addEventListener("keydown", handleKeyNavigation);
    }
    return (): void => {
      document.removeEventListener("mousedown", handleOutsideClick);
      document.removeEventListener("touchstart", handleOutsideClick);
      document.removeEventListener("scroll", hideMenu);
      document.removeEventListener("keydown", handleKeyNavigation);
    };
  }, [
    menuRef,
    hideMenu,
    selectedIndex,
    setSelectedIndex,
    selectables,
    handleElementSelect,
    isVisible,
    hideOnScroll,
  ]);

  useLayoutEffect(() => {
    if (isVisible) {
      const rect = menuRef.current!.getBoundingClientRect();
      const { top, left } = rtl
        ? getRTLMenuPosition(rect, coords)
        : getMenuPosition(rect, coords);
      setStyles((st) => ({
        ...st,
        transform: `translate3d(${left}px, ${top}px, 0)`,
        opacity: 1,
        pointerEvents: "auto",
      }));
    } else {
      setStyles(baseStyles);
    }
  }, [menuRef, isVisible, coords, rtl]);

  const bindMenuProps = {
    style,
    ref: menuRef,
    role: "menu",
    tabIndex: 0,
    onContextMenu: (e: React.SyntheticEvent): void => e.preventDefault(),
    "aria-hidden": !isVisible,
  };
  const bindMenuItemProps = {
    ref: markSelectable,
    role: "menuitem",
    tabIndex: -1,
  };

  return [
    bindMenuProps,
    bindMenuItemProps,
    buildUseContextMenuTrigger(triggerVisible),
    {
      data: collectedData,
      hideMenu,
      isVisible,
      setVisible,
      coords,
      setCoords,
    },
  ];
};

type BindMenuProps = {
  style: {
    position: string;
    opacity: number;
    pointerEvents: string;
    top: number;
    left: number;
    userSelect: string;
    transform: string;
  };
  ref: React.MutableRefObject<HTMLElement | undefined>;
  role: string;
  tabIndex: number;
  onContextMenu: (e: React.SyntheticEvent) => void;
  "aria-hidden": boolean;
};

type BindMenuItemProps = {
  ref: (el: HTMLElement) => HTMLElement[];
  role: string;
  tabIndex: number;
};

type UseContextMenuResult = [
  BindMenuProps,
  BindMenuItemProps,
  (config?: Partial<ContextTriggerConfig>) => UseContextTriggerResult,
  {
    data: unknown;
    hideMenu: () => void;
    isVisible: boolean;
    setVisible: React.Dispatch<React.SetStateAction<boolean>>;
    coords: [number, number];
    setCoords: React.Dispatch<React.SetStateAction<[number, number]>>;
  }
];

export default useContextMenu;
