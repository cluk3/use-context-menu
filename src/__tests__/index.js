import { renderHook, act } from "@testing-library/react-hooks";
import useContextMenu, { keyCodes } from "../useContextMenu";

const getMenuRefMock = () => ({
  getBoundingClientRect: jest.fn(() => ({ height: 100, width: 100 })),
  contains: jest.fn(() => false),
});

test("useContextMenu result is correct", () => {
  const { result } = renderHook(() => useContextMenu());

  expect(result.current.length).toBe(4);

  expect(result.current).toMatchSnapshot();
});

test("useContextMenu registers event listeners only when visible", () => {
  const addEventListener = jest.spyOn(document, "addEventListener");
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  expect(addEventListener).not.toHaveBeenCalled();

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([100, 100]);
  });

  expect(addEventListener).toHaveBeenCalledTimes(3);
  expect(addEventListener.mock.calls.map((call) => call[0])).toEqual([
    "mousedown",
    "touchstart",
    "keydown",
  ]);
});

test("useContextMenu sets correctly the style when changing visibility", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  expect(result.current[0].style).toMatchSnapshot();

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([100, 100]);
  });

  expect(result.current[0].style).toMatchSnapshot();

  act(() => {
    result.current[3].setVisible(false);
  });

  expect(result.current[0].style).toMatchSnapshot();
});

test("useContextMenu keeps the menu inside the viewport", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  expect(result.current[0].style).toMatchSnapshot();

  act(() => {
    result.current[3].setVisible(true);
    result.current[3].setCoords([1000, 700]);
  });

  expect(result.current[0].style).toMatchSnapshot();
});

test("useContextMenu hides the menu when clicking outside", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new MouseEvent("mousedown");

    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu when touching outside", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new MouseEvent("touchstart");

    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on scroll when specified", () => {
  const { result } = renderHook(() => useContextMenu({ hideOnScroll: true }));

  result.current[0].ref.current = getMenuRefMock();

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    document.dispatchEvent(new Event("scroll"));
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on ESCAPE key press", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: keyCodes.ESCAPE });
    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu hides the menu on ENTER key press", () => {
  const { result } = renderHook(() => useContextMenu());

  result.current[0].ref.current = getMenuRefMock();

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: keyCodes.ENTER });
    document.dispatchEvent(event);
  });

  expect(result.current[3].isVisible).toEqual(false);
});

test("useContextMenu index of item selected by keyboard gets reset when closing the menu", () => {
  let selectedEl = 0;
  const { result } = renderHook(() =>
    useContextMenu({
      handleElementSelect: (_selectedEl) => (selectedEl = _selectedEl),
    })
  );

  result.current[0].ref.current = getMenuRefMock();

  result.current[1].ref(1);
  result.current[1].ref(2);
  result.current[1].ref(3);

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: 40 });

    document.dispatchEvent(event);
  });

  act(() => {
    const event = new MouseEvent("touchstart");

    document.dispatchEvent(event);
  });

  act(() => {
    result.current[3].setVisible(true);
  });

  act(() => {
    const event = new KeyboardEvent("keydown", { keyCode: 40 });

    document.dispatchEvent(event);
  });

  expect(selectedEl).toEqual(1);
});

describe("useContextTrigger", () => {
  it("matches snapshot when giving no config", () => {
    const { result } = renderHook(() => useContextMenu());

    const [, , useContextTrigger] = result.current;

    const { result: contextTriggerResult } = renderHook(() =>
      useContextTrigger()
    );

    expect(contextTriggerResult.current).toMatchSnapshot();
  });
  it("matches snapshot when setting mouse holdToDisplay", () => {
    const { result } = renderHook(() => useContextMenu());

    const [, , useContextTrigger] = result.current;

    const { result: contextTriggerResult } = renderHook(() =>
      useContextTrigger({
        holdToDisplay: {
          mouse: 1000,
        },
      })
    );

    expect(contextTriggerResult.current).toMatchSnapshot();
  });
  it("matches snapshot when setting touch holdToDisplay", () => {
    const { result } = renderHook(() => useContextMenu());

    const [, , useContextTrigger] = result.current;

    const { result: contextTriggerResult } = renderHook(() =>
      useContextTrigger({
        holdToDisplay: {
          touch: 1000,
        },
      })
    );

    expect(contextTriggerResult.current).toMatchSnapshot();
  });
});
