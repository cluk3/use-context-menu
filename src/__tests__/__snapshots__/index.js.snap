// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContextMenu keeps the menu inside the viewport 1`] = `
Object {
  "left": 0,
  "opacity": 0,
  "pointerEvents": "none",
  "position": "absolute",
  "top": 0,
  "transform": "translate3d(0,0,0)",
  "userSelect": "none",
}
`;

exports[`useContextMenu keeps the menu inside the viewport 2`] = `
Object {
  "left": 0,
  "opacity": 1,
  "pointerEvents": "auto",
  "position": "absolute",
  "top": 0,
  "transform": "translate3d(900px, 600px, 0)",
  "userSelect": "none",
}
`;

exports[`useContextMenu result is correct 1`] = `
Array [
  Object {
    "aria-hidden": true,
    "onContextMenu": [Function],
    "ref": Object {
      "current": undefined,
    },
    "role": "menu",
    "style": Object {
      "left": 0,
      "opacity": 0,
      "pointerEvents": "none",
      "position": "absolute",
      "top": 0,
      "transform": "translate3d(0,0,0)",
      "userSelect": "none",
    },
    "tabIndex": 0,
  },
  Object {
    "ref": [Function],
    "role": "menuitem",
    "tabIndex": -1,
  },
  [Function],
  Object {
    "coords": Array [
      0,
      0,
    ],
    "data": undefined,
    "hideMenu": [Function],
    "isVisible": false,
    "setCoords": [Function],
    "setVisible": [Function],
  },
]
`;

exports[`useContextMenu sets correctly the style when changing visibility 1`] = `
Object {
  "left": 0,
  "opacity": 0,
  "pointerEvents": "none",
  "position": "absolute",
  "top": 0,
  "transform": "translate3d(0,0,0)",
  "userSelect": "none",
}
`;

exports[`useContextMenu sets correctly the style when changing visibility 2`] = `
Object {
  "left": 0,
  "opacity": 1,
  "pointerEvents": "auto",
  "position": "absolute",
  "top": 0,
  "transform": "translate3d(100px, 100px, 0)",
  "userSelect": "none",
}
`;

exports[`useContextMenu sets correctly the style when changing visibility 3`] = `
Object {
  "left": 0,
  "opacity": 0,
  "pointerEvents": "none",
  "position": "absolute",
  "top": 0,
  "transform": "translate3d(0,0,0)",
  "userSelect": "none",
}
`;

exports[`useContextTrigger matches snapshot when giving no config 1`] = `
Array [
  Object {
    "onContextMenu": [Function],
  },
  [Function],
]
`;

exports[`useContextTrigger matches snapshot when setting mouse holdToDisplay 1`] = `
Array [
  Object {
    "onContextMenu": [Function],
    "onMouseDown": [Function],
    "onMouseOut": [Function],
    "onMouseUp": [Function],
  },
  [Function],
]
`;

exports[`useContextTrigger matches snapshot when setting touch holdToDisplay 1`] = `
Array [
  Object {
    "onContextMenu": [Function],
    "onTouchCancel": [Function],
    "onTouchEnd": [Function],
    "onTouchMove": [Function],
    "onTouchStart": [Function],
  },
  [Function],
]
`;
